<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#1.-Load-data-and-look-into-the-data-information" data-toc-modified-id="1.-Load-data-and-look-into-the-data-information-1">1. Load data and look into the data information</a></span><ul class="toc-item"><li><span><a href="#1.1-show-the-first-5-rows" data-toc-modified-id="1.1-show-the-first-5-rows-1.1">1.1 show the first 5 rows</a></span></li><li><span><a href="#1.2-show-the-dimension-of-the-data" data-toc-modified-id="1.2-show-the-dimension-of-the-data-1.2">1.2 show the dimension of the data</a></span></li><li><span><a href="#1.3-missing-value-judgment" data-toc-modified-id="1.3-missing-value-judgment-1.3">1.3 missing value judgment</a></span></li><li><span><a href="#1.4-show-the-types-of-variables" data-toc-modified-id="1.4-show-the-types-of-variables-1.4">1.4 show the types of variables</a></span></li><li><span><a href="#1.5-show-the-statistical-information-of-variables" data-toc-modified-id="1.5-show-the-statistical-information-of-variables-1.5">1.5 show the statistical information of variables</a></span></li><li><span><a href="#1.6-show-the-correlations-of-variables" data-toc-modified-id="1.6-show-the-correlations-of-variables-1.6">1.6 show the correlations of variables</a></span></li><li><span><a href="#1.7-show-the-number-of-target-variable" data-toc-modified-id="1.7-show-the-number-of-target-variable-1.7">1.7 show the number of target variable</a></span></li></ul></li><li><span><a href="#2.-Model-training" data-toc-modified-id="2.-Model-training-2">2. Model training</a></span><ul class="toc-item"><li><span><a href="#2.1-Data-prepare" data-toc-modified-id="2.1-Data-prepare-2.1">2.1 Data prepare</a></span><ul class="toc-item"><li><span><a href="#2.1.1-separate-the-variables-and-target" data-toc-modified-id="2.1.1-separate-the-variables-and-target-2.1.1">2.1.1 separate the variables and target</a></span></li><li><span><a href="#2.1.2-standerdize-the-variables" data-toc-modified-id="2.1.2-standerdize-the-variables-2.1.2">2.1.2 standerdize the variables</a></span></li><li><span><a href="#2.1.3-encoder-the-target" data-toc-modified-id="2.1.3-encoder-the-target-2.1.3">2.1.3 encoder the target</a></span></li><li><span><a href="#2.1.4-separate-the-train-and-test-data-set" data-toc-modified-id="2.1.4-separate-the-train-and-test-data-set-2.1.4">2.1.4 separate the train and test data set</a></span></li></ul></li><li><span><a href="#2.2-Algorithm-prepare" data-toc-modified-id="2.2-Algorithm-prepare-2.2">2.2 Algorithm prepare</a></span></li><li><span><a href="#2.3-Model-training" data-toc-modified-id="2.3-Model-training-2.3">2.3 Model training</a></span></li></ul></li><li><span><a href="#3.-Conclusion" data-toc-modified-id="3.-Conclusion-3">3. Conclusion</a></span></li></ul></div>

In [1]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder, StandardScaler # preprocessing the X and y
from sklearn.model_selection import train_test_split # separate the train and test data set
from sklearn import metrics
import xgboost as xgb 

refer: https://xgboost.readthedocs.io/en/latest/python/python_intro.html

## 1. Load data and look into the data information

In [2]:
df=pd.read_csv('./voice.csv')

### 1.1 show the first 5 rows

In [3]:
df.head(5)

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

### 1.2 show the dimension of the data

In [4]:
df.shape

(3168, 21)

### 1.3 missing value judgment

In [5]:
df.isnull().sum()

meanfreq    0
sd          0
median      0
Q25         0
Q75         0
IQR         0
skew        0
kurt        0
sp.ent      0
sfm         0
mode        0
centroid    0
meanfun     0
minfun      0
maxfun      0
meandom     0
mindom      0
maxdom      0
dfrange     0
modindx     0
label       0
dtype: int64

### 1.4 show the types of variables

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3168 entries, 0 to 3167
Data columns (total 21 columns):
meanfreq    3168 non-null float64
sd          3168 non-null float64
median      3168 non-null float64
Q25         3168 non-null float64
Q75         3168 non-null float64
IQR         3168 non-null float64
skew        3168 non-null float64
kurt        3168 non-null float64
sp.ent      3168 non-null float64
sfm         3168 non-null float64
mode        3168 non-null float64
centroid    3168 non-null float64
meanfun     3168 non-null float64
minfun      3168 non-null float64
maxfun      3168 non-null float64
meandom     3168 non-null float64
mindom      3168 non-null float64
maxdom      3168 non-null float64
dfrange     3168 non-null float64
modindx     3168 non-null float64
label       3168 non-null object
dtypes: float64(20), object(1)
memory usage: 519.9+ KB


### 1.5 show the statistical information of variables

In [7]:
df.describe()

meanfreq           sd       median          Q25          Q75  \
count  3168.000000  3168.000000  3168.000000  3168.000000  3168.000000   
mean      0.180907     0.057126     0.185621     0.140456     0.224765   
std       0.029918     0.016652     0.036360     0.048680     0.023639   
min       0.039363     0.018363     0.010975     0.000229     0.042946   
25%       0.163662     0.041954     0.169593     0.111087     0.208747   
50%       0.184838     0.059155     0.190032     0.140286     0.225684   
75%       0.199146     0.067020     0.210618     0.175939     0.243660   
max       0.251124     0.115273     0.261224     0.247347     0.273469   

               IQR         skew         kurt       sp.ent          sfm  \
count  3168.000000  3168.000000  3168.000000  3168.000000  3168.000000   
mean      0.084309     3.140168    36.568461     0.895127     0.408216   
std       0.042783     4.240529   134.928661     0.044980     0.177521   
min       0.014558     0.141735     2.068455     0.738651     0.036876   
25%       0.042560     1.649569     5.669547     0.861811     0.258041   
50%       0.094280     2.197101     8.318463     0.901767     0.396335   
75%       0.114175     2.931694    13.648905     0.928713     0.533676   
max       0.252225    34.725453  1309.612887     0.981997     0.842936   

              mode     centroid      meanfun       minfun       maxfun  \
count  3168.000000  3168.000000  3168.000000  3168.000000  3168.000000   
mean      0.165282     0.180907     0.142807     0.036802     0.258842   
std       0.077203     0.029918     0.032304     0.019220     0.030077   
min       0.000000     0.039363     0.055565     0.009775     0.103093   
25%       0.118016     0.163662     0.116998     0.018223     0.253968   
50%       0.186599     0.184838     0.140519     0.046110     0.271186   
75%       0.221104     0.199146     0.169581     0.047904     0.277457   
max       0.280000     0.251124     0.237636     0.204082     0.279114   

           meandom       mindom       maxdom      dfrange      modindx  
count  3168.000000  3168.000000  3168.000000  3168.000000  3168.000000  
mean      0.829211     0.052647     5.047277     4.994630     0.173752  
std       0.525205     0.063299     3.521157     3.520039     0.119454  
min       0.007812     0.004883     0.007812     0.000000     0.000000  
25%       0.419828     0.007812     2.070312     2.044922     0.099766  
50%       0.765795     0.023438     4.992188     4.945312     0.139357  
75%       1.177166     0.070312     7.007812     6.992188     0.209183  
max       2.957682     0.458984    21.867188    21.843750     0.932374

### 1.6 show the correlations of variables

In [8]:
df.corr()

meanfreq        sd    median       Q25       Q75       IQR  \
meanfreq  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
sd       -0.739039  1.000000 -0.562603 -0.846931 -0.161076  0.874660   
median    0.925445 -0.562603  1.000000  0.774922  0.731849 -0.477352   
Q25       0.911416 -0.846931  0.774922  1.000000  0.477140 -0.874189   
Q75       0.740997 -0.161076  0.731849  0.477140  1.000000  0.009636   
IQR      -0.627605  0.874660 -0.477352 -0.874189  0.009636  1.000000   
skew     -0.322327  0.314597 -0.257407 -0.319475 -0.206339  0.249497   
kurt     -0.316036  0.346241 -0.243382 -0.350182 -0.148881  0.316185   
sp.ent   -0.601203  0.716620 -0.502005 -0.648126 -0.174905  0.640813   
sfm      -0.784332  0.838086 -0.661690 -0.766875 -0.378198  0.663601   
mode      0.687715 -0.529150  0.677433  0.591277  0.486857 -0.403764   
centroid  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
meanfun   0.460844 -0.466281  0.414909  0.545035  0.155091 -0.534462   
minfun    0.383937 -0.345609  0.337602  0.320994  0.258002 -0.222680   
maxfun    0.274004 -0.129662  0.251328  0.199841  0.285584 -0.069588   
meandom   0.536666 -0.482726  0.455943  0.467403  0.359181 -0.333362   
mindom    0.229261 -0.357667  0.191169  0.302255 -0.023750 -0.357037   
maxdom    0.519528 -0.482278  0.438919  0.459683  0.335114 -0.337877   
dfrange   0.515570 -0.475999  0.435621  0.454394  0.335648 -0.331563   
modindx  -0.216979  0.122660 -0.213298 -0.141377 -0.216475  0.041252   

              skew      kurt    sp.ent       sfm      mode  centroid  \
meanfreq -0.322327 -0.316036 -0.601203 -0.784332  0.687715  1.000000   
sd        0.314597  0.346241  0.716620  0.838086 -0.529150 -0.739039   
median   -0.257407 -0.243382 -0.502005 -0.661690  0.677433  0.925445   
Q25      -0.319475 -0.350182 -0.648126 -0.766875  0.591277  0.911416   
Q75      -0.206339 -0.148881 -0.174905 -0.378198  0.486857  0.740997   
IQR       0.249497  0.316185  0.640813  0.663601 -0.403764 -0.627605   
skew      1.000000  0.977020 -0.195459  0.079694 -0.434859 -0.322327   
kurt      0.977020  1.000000 -0.127644  0.109884 -0.406722 -0.316036   
sp.ent   -0.195459 -0.127644  1.000000  0.866411 -0.325298 -0.601203   
sfm       0.079694  0.109884  0.866411  1.000000 -0.485913 -0.784332   
mode     -0.434859 -0.406722 -0.325298 -0.485913  1.000000  0.687715   
centroid -0.322327 -0.316036 -0.601203 -0.784332  0.687715  1.000000   
meanfun  -0.167668 -0.194560 -0.513194 -0.421066  0.324771  0.460844   
minfun   -0.216954 -0.203201 -0.305826 -0.362100  0.385467  0.383937   
maxfun   -0.080861 -0.045667 -0.120738 -0.192369  0.172329  0.274004   
meandom  -0.336848 -0.303234 -0.293562 -0.428442  0.491479  0.536666   
mindom   -0.061608 -0.103313 -0.294869 -0.289593  0.198150  0.229261   
maxdom   -0.305651 -0.274500 -0.324253 -0.436649  0.477187  0.519528   
dfrange  -0.304640 -0.272729 -0.319054 -0.431580  0.473775  0.515570   
modindx  -0.169325 -0.205539  0.198074  0.211477 -0.182344 -0.216979   

           meanfun    minfun    maxfun   meandom    mindom    maxdom  \
meanfreq  0.460844  0.383937  0.274004  0.536666  0.229261  0.519528   
sd       -0.466281 -0.345609 -0.129662 -0.482726 -0.357667 -0.482278   
median    0.414909  0.337602  0.251328  0.455943  0.191169  0.438919   
Q25       0.545035  0.320994  0.199841  0.467403  0.302255  0.459683   
Q75       0.155091  0.258002  0.285584  0.359181 -0.023750  0.335114   
IQR      -0.534462 -0.222680 -0.069588 -0.333362 -0.357037 -0.337877   
skew     -0.167668 -0.216954 -0.080861 -0.336848 -0.061608 -0.305651   
kurt     -0.194560 -0.203201 -0.045667 -0.303234 -0.103313 -0.274500   
sp.ent   -0.513194 -0.305826 -0.120738 -0.293562 -0.294869 -0.324253   
sfm      -0.421066 -0.362100 -0.192369 -0.428442 -0.289593 -0.436649   
mode      0.324771  0.385467  0.172329  0.491479  0.198150  0.477187   
centroid  0.460844  0.383937  0.274004  0.536666  0.229261  0.519528   
meanfun   1.000000  0.339387  0.311950  0.270840

### 1.7 show the number of target variable

In [9]:
print('total number    ：{}'.format(df.shape[0]))
print('number of male  ：{}'.format(df[df.label=='male'].shape[0]))
print('number of female：{}'.format(df[df.label=='female'].shape[0]))

total number    ：3168
number of male  ：1584
number of female：1584


## 2. Model training

### 2.1 Data prepare
#### 2.1.1 separate the variables and target

In [10]:
X = df.iloc[:, :-1] 
y = df.iloc[:, -1]
# [:,:-1]第一个位置为行，取全部行；
# 第二个位置为列，从第一列取到倒数第二列

In [11]:
print("show the variables: \n",X.head())

show the variables: 
    meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm      mode  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  0.000000  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  0.000000  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  0.000000  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  0.083878  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  0.104261  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange 

In [12]:
print("show the target: \n",y.head())

show the target: 
 0    male
1    male
2    male
3    male
4    male
Name: label, dtype: object


#### 2.1.2 standerdize the variables

In [13]:
scaler = StandardScaler()
X = scaler.fit_transform(X)
print(X)

[[-4.04924806  0.4273553  -4.22490077 ... -1.43142165 -1.41913712
  -1.45477229]
 [-3.84105325  0.6116695  -3.99929342 ... -1.41810716 -1.4058184
  -1.01410294]
 [-3.46306647  1.60384791 -4.09585052 ... -1.42920257 -1.41691733
  -1.06534356]
 ...
 [-1.29877326  2.32272355 -0.05197279 ... -0.5992661  -0.58671739
   0.17588664]
 [-1.2452018   2.012196   -0.01772849 ... -0.41286326 -0.40025537
   1.14916112]
 [-0.51474626  2.14765111 -0.07087873 ... -1.27608595 -1.2637521
   1.47567886]]


#### 2.1.3 encoder the target

In [14]:
gender_encoder = LabelEncoder()
y = gender_encoder.fit_transform(y)
print(y)

[1 1 1 ... 0 0 0]


#### 2.1.4 separate the train and test data set

In [15]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

train_data = xgb.DMatrix(X_train, label=y_train)
test_data = xgb.DMatrix(X_test, label=y_test)

### 2.2 Algorithm prepare

In [16]:
param = {'boosting_type':'gbdt',
                         'objective' : 'binary:logistic', # 任务目标
                         'eval_metric' : 'auc',           # 评估指标
                         'eta' : 0.01,                    # 学习率
                         'max_depth' : 15,                # 树最大深度
                         'colsample_bytree':0.8,          # 设置在每次迭代中使用特征的比例
                         'subsample': 0.9,                # 样本采样比例
                         'subsample_freq': 8,             # boosting的次数
                         'alpha': 0.6,                    # L1正则
                         'lambda': 0,                     # L2正则
        }

### 2.3 Model training

xgboost.train(params, dtrain, num_boost_round=10, evals=(), obj=None, feval=None, maximize=False, early_stopping_rounds=None, evals_result=None, verbose_eval=True, xgb_model=None, callbacks=None)

Parameters
- params (dict) – Booster params.
- dtrain (DMatrix) – Data to be trained.
- num_boost_round (int) – Number of boosting iterations.
- evals (list of pairs (DMatrix, string)) – List of validation sets for which metrics will evaluated during training. Validation metrics will help us track the performance of the model.

In [17]:
model = xgb.train(param, 
                  train_data, 
                  evals=[(train_data, 'train'), (test_data, 'test')], 
                  num_boost_round = 5000, 
                  early_stopping_rounds=20, 
                  verbose_eval=25)

[23:35:58] WARNING: C:\Users\Administrator\workspace\xgboost-win64_release_1.2.0\src\learner.cc:516: 
Parameters: { boosting_type, subsample_freq } might not be used.

  This may not be accurate due to some parameters are only used in language bindings but
  passed down to XGBoost core.  Or some parameters are not used but slip through this
  verification. Please open an issue if you find above cases.


[0]	train-auc:0.99065	test-auc:0.99292
Multiple eval metrics have been passed: 'test-auc' will be used for early stopping.

Will train until test-auc hasn't improved in 20 rounds.
[25]	train-auc:0.99945	test-auc:0.99782
Stopping. Best iteration:
[21]	train-auc:0.99948	test-auc:0.99783



**The model shows, the model stop training after 25th iteration, and the final test-auc is 0.99782.  
Model training has random part, so you'd better set a seed.**

## 3. Conclusion

- 自己还不能很自如的用一些package，比如可以在前面corr地方用热力图展现，后面xgboosting的结果是否也可以用图像展现
- 收获就是学会了看官方文档，但是我对整个package框架如何搭建不是很熟悉，比如class等概念不熟悉，所以看起来很耗费时间。所以当时间充足的时候，我可以边看官方文档，边把代码改写出来，也可以主动多看一些相关的技术文档。但是时间不充足的时候，就只能把老师的代码写一遍，然后添加上自己的注释和思考
- 虽然会看了官方文档，但是还没学会看源代码，这点有待学习
- 下次一定早一点完成作业，这样有更多的时间可以优化